### PR TITLE
fix(transcript): resolve speaker timeline color as of the meeting date

### LINKED ISSUE #309: Independent councilor color not updating in speaker timeline
### Concept

Dimitris Tselempis from Orestiada's council became independent, and his party affiliation has been correctly updated in the database. However, the speaker timeline bar still displays his old party's color instead of the default independent color.

This is a display logic issue — the data is correct, but the timeline is not reflecting the updated affiliation.

### User Story

As a citizen viewing a council meeting, I want the speaker timeline to show the correct party color for each councilor, so that I can accurately understand their current political affiliation.

### Context

- His color should update for all meetings dated **after November 1, 2025**. Meetings before that date should retain his old party color.
- The platform already has a default color for independent councilors.
- Currently only affects Dimitris Tselempis in Orestiada — no other councilors in this municipality have the same issue.
- The speaker timeline likely resolves party color without considering when the affiliation changed, or it may be caching the old value.

### PR DESCRIPTION
**Pull request by VasiaKoum**
Fixes #309

## Problem

The speaker timeline resolved each councilor's party color with `getPartyFromRoles(person.roles)` — **without a date** — so it always used the *current* affiliation. After Dimitris Tselempis (Orestiada) became independent, his old party color no longer matched the affiliation valid at the time of each meeting.

The date-aware logic already existed (`getPartyFromRoles(roles, date)` + `isRoleActiveAt`); the UI call sites simply weren't passing the meeting date.

## Fix

Thread the meeting date into the in-meeting speaker timeline surfaces:

- **`PersonBadge`** — add an optional, backwards-compatible `date?: Date` prop, used for party-color resolution in `PersonDisplay` (avatar/initials), the `inline` variant (party dot), and the people-picker list. Callers that omit it keep today's behavior.
- **`SpeakerSegment`** — pass `meeting.dateTime` to the segment's colored left border and both `PersonBadge` renders.
- **`SpeakersOverviewSheet`** — pass `meeting.dateTime` to the per-meeting speaker stats.

## Historical accuracy preserved

`isRoleActiveAt` treats `endDate` as **exclusive** (`end > date`):

- Meeting **before** the affiliation ended → old party role active → **old color preserved**.
- Meeting **on/after** the affiliation ended → role inactive → falls back to the default independent color.

Added `getPartyFromRoles` unit tests covering the before / after / on-the-boundary cases.

## Scope

Display-only fix; relies on the DB already holding the correct role `endDate` (per the issue, the data is already correct). Search results, person profile pages, and subject cards are intentionally untouched — those show *current* affiliation and are outside the timeline.

## Verification

- Role tests: 46 passed (3 new)
- Full suite: 834 passed
- `npx tsc --noEmit`: clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

This PR fixes the speaker timeline party-color resolution so it uses the affiliation valid at the time of each meeting rather than the current affiliation. The root serialization bug (ISO-string `meeting.dateTime` causing `NaN` comparisons) was caught in a prior review round and fixed centrally in `isRoleActiveAt` by normalizing the `date` parameter with `new Date(date)`.

- **`isRoleActiveAt`** now wraps `date` in `new Date(...)` so ISO strings serialized during Server→Client Component prop passing compare correctly against role dates.
- **`getPartyFromRoles`** call sites in `TranscriptControls`, `SpeakerSegment`, and `SpeakersOverviewSheet` are updated to pass `meeting.dateTime`, and `PersonBadge` gains an optional backwards-compatible `date` prop threaded through to `PersonDisplay`.
- **Tests** cover the before/after/on-boundary cases and a dedicated ISO-string regression test for the serialization fix.

<h3>Confidence Score: 5/5</h3>

Display-only fix with no data mutations; the change is isolated to party-color resolution in the transcript UI.

The serialization issue raised in prior review rounds is fixed centrally and covered by a new regression test. All six changed files make narrow, targeted adjustments with clear backwards compatibility; the only behavioral difference is that older meetings now show the historically-correct party color instead of the current one.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/lib/utils/roles.ts | Adds `const at = new Date(date)` normalization in `isRoleActiveAt` so ISO strings from Next.js Client Component prop serialization are handled correctly — the central fix. |
| src/lib/utils/__tests__/roles.test.ts | Adds three boundary tests for `getPartyFromRoles` (before/after/on-endDate) plus an ISO-string regression test that guards the serialization fix. |
| src/components/meetings/TranscriptControls.tsx | Threads `meeting.dateTime` into three `getPartyFromRoles` call sites (hover tooltip, current speaker, segment timeline strip) in place of the naïve first-party-role lookup. |
| src/components/meetings/transcript/SpeakerSegment.tsx | Passes `meetingDate` to the border-color calculation and both `PersonBadge` renders (inline header and people-picker). |
| src/components/meetings/transcript/SpeakersOverviewSheet.tsx | Passes `meeting.dateTime` to `getPartyFromRoles` and adds it to the `useMemo` dependency array so the stats recompute if the meeting object changes. |
| src/components/persons/PersonBadge.tsx | Adds optional backwards-compatible `date?: Date` prop threaded through `PersonDisplayProps` → `PersonBadgeProps` → `PersonDisplay` and the `inline` variant — no breaking change to existing callers. |

</details>

<sub>Reviews (3): Last reviewed commit: ["fix(transcript): resolve speaker timelin..."](https://github.com/schemalabz/opencouncil/commit/f670e2ab777cfc2c5e2eabcd5ba7260401fd10eb) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=38353658)</sub>

### CODE REVIEW

File: src/components/meetings/transcript/SpeakerSegment.tsx [hunk 1]
```diff
@@ -174,7 +174,10 @@ const SpeakerSegment = React.memo(({ segment, isFirstSegment }: {
 }) => {
     // useCouncilMeetingMeta() — not useCouncilMeetingData() — so this
     // component bails on transcript-only edits.
-    const { getPerson, getSpeakerTag, getSpeakerSegmentCount, people, speakerTags } = useCouncilMeetingMeta();
+    const { getPerson, getSpeakerTag, getSpeakerSegmentCount, people, speakerTags, meeting } = useCouncilMeetingMeta();
+    // Resolve party color as of the meeting date so a councilor who later
+    // changed affiliation keeps their historical color on older meetings.
+    const meetingDate = meeting.dateTime;
```
[the remark this answers is not in the document]
Comment (VasiaKoum) [reply src/components/meetings/transcript/SpeakerSegment.tsx:180]: Good catch — confirmed: `meeting.dateTime` arrives as an ISO string in the Client Component, and `Date <= string` / `Date > string` coerce to `NaN`, so every dated role would appear inactive (dropping the color for all councilors, not just independents).

Fixed at the root instead of the call site: `isRoleActiveAt` already normalizes the role's `start`/`end` via `new Date(...)` for this exact JSON-serialization reason, so I normalized the `date` parameter there too (`const at = new Date(date)`). This repairs both this call site and `SpeakersOverviewSheet`, plus any future caller, without scattering `new Date()` wraps. Added a `getPartyFromRoles` regression test that passes a serialized ISO-string date.

File: src/components/meetings/transcript/SpeakersOverviewSheet.tsx [hunk 2]
```diff
@@ -48,7 +48,7 @@ function SpeakerStatsContent() {
 
             if (!stats.has(key)) {
                 const person = personId ? getPerson(personId) : null;
-                const party = person ? getPartyFromRoles(person.roles) : null;
+                const party = person ? getPartyFromRoles(person.roles, meeting.dateTime) : null;
```
[the remark this answers is not in the document]
Comment (VasiaKoum) [reply src/components/meetings/transcript/SpeakersOverviewSheet.tsx:51]: Same root cause as the `SpeakerSegment` thread — fixed centrally in `isRoleActiveAt` by normalizing the `date` parameter, so this call site (passing `meeting.dateTime` directly) now resolves correctly. Covered by the new serialized-string regression test.

**Comment by VasiaKoum** [comment]
I [force-pushed](https://github.com/schemalabz/opencouncil/compare/0f049fab31156992138eefd02a0e578887be5184..24c8ea37db6114f03c62a3f29238c9ae25558174) to address Greptile's two P1 comments (both the same serialization issue):

- `meeting.dateTime` is serialized to an ISO string in the Client Component. `isRoleActiveAt` normalized the role's `start`/`end` but compared them against the raw `date` param — `Date`-vs-`string` comparisons coerce to `NaN`, so every dated role appeared inactive and the color dropped to the default for **all** councilors, not just independents.
- Fixed at the source: `isRoleActiveAt` now normalizes the `date` parameter (`const at = new Date(date)`), mirroring how it already normalizes role dates. This fixes both call sites (`SpeakerSegment`, `SpeakersOverviewSheet`) and any future caller.
- Added a `getPartyFromRoles` regression test that passes a serialized ISO-string date (before/after the affiliation change).

Full suite: 835 passed. `tsc --noEmit`: clean. Both review threads resolved.

**Comment by VasiaKoum** [comment]
I [force-pushed](https://github.com/schemalabz/opencouncil/compare/24c8ea37db6114f03c62a3f29238c9ae25558174..f670e2ab777cfc2c5e2eabcd5ba7260401fd10eb) to cover the **player's colored seek bar**, which was the surface still showing the old color.

`TranscriptControls.tsx` (the meeting media player) resolved each speaker segment's color with a date-blind inline `person?.roles?.find(role => role.party)?.party` — it never went through `getPartyFromRoles` and ignored the meeting date entirely. Switched all three sites (the seek-bar segments, the hovered-speaker tooltip, and the current-speaker indicator) to `getPartyFromRoles(person.roles, meeting.dateTime)`.

That was the missing piece — the earlier commit fixed the transcript view and overview, but not the player bar.

`tsc --noEmit`: clean. Full suite: 835 passed.

Could you confirm on the PR preview (pr-484) that Tselempis now shows the independent color on post-Nov-2025 meetings and his old party color on earlier ones? I can't reproduce locally since the development DB doesn't have that councilor's data.

**Review by christosporios — APPROVED** [review]
